### PR TITLE
ingressClass support

### PR DESCRIPTION
**Pull request by bshifter**
This PR extends the charts with the possibility of using ingressClassName for ingress resources instead the deprecated ingress.classname annotation.

### CODE REVIEW

File: charts/dex/values.yaml [hunk 1]
```diff
@@ -120,6 +120,8 @@ ingress:
   #  - secretName: chart-example-tls
   #    hosts:
   #      - chart-example.local
+  # ingressClassName support
```
Comment (nabokihms) [inline charts/dex/values.yaml:123]: Could you add more info about what this parameter does (maybe with the link to Kubernetes documentation)?

**Review by nabokihms — COMMENTED** [review]
Hello, and thank you for the PR! I added some suggestions. Please look through when you have time.

File: charts/dex/templates/ingress.yaml [hunk 2]
```diff
@@ -16,6 +16,9 @@ metadata:
     {{- toYaml . | nindent 4 }}
   {{- end }}
 spec:
+  {{- if .Values.ingress.className }}
+  ingressClassName: {{ .Values.ingress.className }}
```
[the remark this answers is not in the document]
Comment (nabokihms) [reply charts/dex/templates/ingress.yaml:20]: Oh, doc even says that you can use `ingressClassName` only starting from Kubernetes >=1.18.

File: charts/dex/templates/ingress.yaml [hunk 3]
```diff
@@ -16,6 +16,12 @@ metadata:
     {{- toYaml . | nindent 4 }}
   {{- end }}
 spec:
+  {{- if .Values.ingress.className }}
+  {{- if not (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion) -}}
+  {{- fail "Kubernetes version >=1.18 required to use className!" -}}
```
Comment (nabokihms) [inline charts/dex/templates/ingress.yaml:21]:
```suggestion
  {{- fail "Kubernetes version >=1.18 required to use ingress.className!" -}}
```

File: charts/dex/values.yaml [hunk 4]
```diff
@@ -120,6 +120,9 @@ ingress:
   #  - secretName: chart-example-tls
   #    hosts:
   #      - chart-example.local
+# Sets Ingress/ingressClassName. This way ingress resources are able to bound specific ingress-controllers. Kubernetes version >=1.18 required.
```
Comment (nabokihms) [inline charts/dex/values.yaml:123]:
```suggestion
  # Sets Ingress/ingressClassName. This way ingress resources are able to bound specific ingress-controllers. Kubernetes version >=1.18 required.
```

File: charts/dex/values.yaml [hunk 5]
```diff
@@ -120,6 +120,9 @@ ingress:
...
+# See [Ingress documentation](https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-class) for details.
```
Comment (nabokihms) [inline charts/dex/values.yaml:124]:
```suggestion
  # See [Ingress documentation](https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-class) for details.
```

**Review by nabokihms — COMMENTED** [review]
Thanks. I left some more minor suggestions. It looks like we are almost done.

File: charts/dex/values.yaml (hunk 4 above)
Comment (sagikazarmark) [reply charts/dex/values.yaml]: Actually, it also needs the `# -- ` part to show up in the generated documentation. Please also add a new line before this.

**Review by sagikazarmark — COMMENTED** [review]
Sorry for joining the party late.

How about we don't limit the ingress class to Kubernetes 1.18, but instead fall back to the old annotation solution (if the annotation is not specified)?

That sounds like a more future proof solution to me, since the annotation will eventually be dropped, but the ingress class won't, so people can start using a future proof solution today.

File: charts/dex/templates/ingress.yaml [hunk 6]
```diff
@@ -11,11 +11,19 @@ metadata:
   name: {{ $fullName }}
   labels:
     {{- include "dex.labels" . | nindent 4 }}
+  {{- if and (.Values.ingress.className) (not (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion)) }}
```
Comment (sagikazarmark) [inline charts/dex/templates/ingress.yaml:14]:
```suggestion
  {{- if and .Values.ingress.className (not (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion)) }}
```

File: charts/dex/templates/ingress.yaml [hunk 7]
```diff
@@ -11,11 +11,19 @@ metadata:
...
+    {{- if not (hasKey .Values.ingress.annotations "kubernetes.io/ingress.class") }}
+    {{- $_ := set .Values.ingress.annotations "kubernetes.io/ingress.class" .Values.ingress.className}}
+    {{- end }}
+  {{- end }}
   {{- with .Values.ingress.annotations }}
   annotations:
     {{- toYaml . | nindent 4 }}
   {{- end }}
 spec:
+  {{- if and (.Values.ingress.className) (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion) }}
```
Comment (sagikazarmark) [inline charts/dex/templates/ingress.yaml:24]:
```suggestion
  {{- if and .Values.ingress.className (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion) }}
```

File: charts/dex/values.yaml [hunk 8]
```diff
@@ -120,6 +120,9 @@ ingress:
   #  - secretName: chart-example-tls
   #    hosts:
   #      - chart-example.local
+  # Sets Ingress/ingressClassName. This way ingress resources are able to bound specific ingress-controllers. Kubernetes version >=1.18 required.
```
Comment (sagikazarmark) [inline charts/dex/values.yaml:123]:
```suggestion

  # -- Sets Ingress/ingressClassName. This way ingress resources are able to bound specific ingress-controllers. Kubernetes version >=1.18 required.
```

**Review by sagikazarmark — CHANGES_REQUESTED** [review]
Couple small changes left, other than that LGTM.

**Review by sagikazarmark — APPROVED** [review]

**Comment by sagikazarmark** [comment]
Thanks!

**Pull request merged**